### ISSUE TITLE
Changing choices configuration requires deleting system

### ISSUE THREAD

**Issue by hazmat345** [beer-garden/beer-garden#58]
Currently if you change only a `choices` definition you need to delete the plugin in order for the changes to be picked up. This is less than ideal.

**Comment by hazmat345** [comment]
@loganasherjones what do we want the behavior for this to be?

Do we want to say that choices definitions can't change unless the plugin is a dev version? That seems like it would be pretty restrictive.

**Comment by hazmat345** [comment]
So looking into this a bit, I think we need to be a little more flexible. On the server side we're currently doing this:

```python
if system.has_different_commands(new_commands):
    if system.commands and 'dev' not in system.version:
        raise ModelValidationError('System %s-%s already exists with '
                                   'different commands' %
                                   (system.name, system.version))
    else:
        system.upsert_commands(new_commands)
```

That has the effect of a catch 22 for non-dev versions: Commands need to be different to be updated, but different commands aren't allowed.

I think we should move to this:

```python
if (
    system.commands
    and 'dev' not in system.version
    and system.has_different_commands(new_commands)
):
    raise ModelValidationError(
        'System %s-%s already exists with different commands' %
        (system.name, system.version)
    )

system.upsert_commands(new_commands)
```

This way we explicitly define what changes and conditions will trigger an error and persist all changes that wouldn't.

This should also have a side benefit of allowing updates to other command fields (like `description`) like we currently allow on systems.

**Comment by loganasherjones** [comment]
I think the new way you're proposing does make sense. I think the big question is: should different choices result in a `True` being returned from `has_different_commands`

**Comment by hazmat345** [comment]
Right, that's the tricky part. I'm pretty torn - I'm trying to look at the worst case for each option.

Worst case for allowing choices changes: A choice that was previously valid is removed and breaks a downstream plugin.

Worst case for not allowing: Populating choices from a URL you don't control, and that URL changes. Now you have to release a 'new' version of the plugin even though nothing has really changed.

I guess I'm leaning towards not allowing changes for non-develop versions? It seems like choices are part of the command API, and we shouldn't allow API changes without version changes. But I could probably be convinced otherwise.

**Comment by hazmat345** [comment]
Thinking about this some more:

If you say `choices=['a', 'b', 'c']` and then later want to change to `choices=['1', '2', '3']` that *feels* like a 'change' to me. Which makes me think we shouldn't allow it.

But if you say `choices=return_abc()` and then `return_abc` changes to actually return `['1', '2', '3']` that *feels* OK to me (we'd also have no way to prevent that anyway). It feels OK even though to a downstream plugin (my main concern above) the effect is the same.

I guess my main issue now is that even if we prevent the changes to choices that we can (aka the first example in this comment) downstream plugins still can't depend on the choices being invariant.

So now I think we should allow non-static choices (url & command-based) to change. I'm still not sure about static choices. If we disallow them changing then downstream plugins could look at the actual parameter definition and know that 'static' means 'these will never change for this version' and anything else means 'these could change, beware.'

Or we could just say 'parameter choices can change.'

**Comment by hazmat345** [comment]
@loganasherjones can you weigh in on this?

**Comment by loganasherjones** [comment]
I would say we want to allow people to change choices without forcing them to release a new version.

The versioning schema that we are enforcing doesn't actually have any weight behind it. Meaning, someone can change the actual underlying implementation of a plugin without changing the commands/parameters and we currently have no way of knowing that. 

If we wanted to go down the path of forcing someone to bump their versions when they change choices,  we must come up with a decent way to actually enforce _any changes_ in a plugin resulting in the force of version bump. Which is mostly out of scope. Either way, we would need #20 to be done before I would feel comfortable even enforcing that.

Bottom line: I'd say allow people to change their choices without bumping versions.
